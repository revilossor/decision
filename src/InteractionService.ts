import Enquirer from 'enquirer';

export class InteractionService {
  static say (message: string) {
    console.log(`\u001b[1m- ${message}\u001b[0m`);
  }

  static async textPrompt (copy: string): Promise<string> {
    const { thing } = await Enquirer.prompt({
      type: 'input',
      name: 'thing',
      message: copy
    });
    return thing;
  }

  static async selectPrompt (copy: string, values: string[]):Promise<string> {
    const { value } = await Enquirer.prompt({
      type: 'select',
      name: 'value',
      message: copy,
      choices: values
    });
    return value;
  }
}
