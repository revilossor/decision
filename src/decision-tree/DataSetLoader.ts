import * as fs from 'fs';
import * as path from 'path';

export class DataSetLoader {
  private static loadFile (filepath: string): Buffer {
    try {
      return fs.readFileSync(path.resolve(filepath));
    } catch (e) {
      throw Error(`there was a problem reading the file: ${e.message}`);
    }
  }

  private static parseBuffer (buffer: Buffer): string[][] {
    try {
      if (buffer.length === 0) {
        throw Error('no content');
      }

      const content = buffer.toString();
      const [attributesRow, ...recordRows] = content.split('\n');

      const attributes = attributesRow.split(',');

      const records = recordRows
        .map(row => row.split(','))
        .filter(row => row.length !== 1 && row[0] !== ''); // allow empty but trim

      if (records.length === 0) {
        throw Error('no records');
      }

      records.forEach(row => {
        if (row.length < attributes.length) {
          throw Error('missing attribute value');
        } else if (row.length > attributes.length) {
          throw Error('extra attribute value');
        }
      });

      return [attributes, ...records];
    } catch (e) {
      throw Error(`there was a problem parsing the file: ${e.message}`);
    }
  }

  public static load (filepath: string): string[][] {
    const buffer = DataSetLoader.loadFile(filepath);
    return DataSetLoader.parseBuffer(buffer);
  }
}
