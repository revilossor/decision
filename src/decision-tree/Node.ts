export class Node {
  public children: Array<Node>

  public constructor (public attribute:string, public value:string) {
    this.children = [];
  }

  public addChild (child: Node): Node {
    this.children.push(child);
    return this;
  }

  public get length ():number { return this.children.length; }
}
