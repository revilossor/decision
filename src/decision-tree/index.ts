export * from './DecisionTree';
export * from './Node';
