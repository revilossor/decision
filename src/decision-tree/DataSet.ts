import { DataSetLoader } from './DataSetLoader';

export class DataSet {
  public records: string[][]
  public attributes: string[]

  constructor (attributes: string[], ...records: string[][]) {
    const lower = (attribute: string) => attribute.toLowerCase();
    this.attributes = attributes.map(lower);
    this.records = records.map(record => record.map(lower));
  }

  public has (attribute: string):boolean {
    return this.attributes.includes(attribute.toLowerCase());
  }

  public subset (attribute: string, value: string): DataSet {
    const index = this.getAttributeIndex(attribute);
    const pertinent = this.records.filter(record => record[index] === value.toLowerCase());
    return new DataSet(this.attributes, ...pertinent);
  }

  private getAttributeIndex (attribute:string): number {
    const key = attribute.toLowerCase();
    const index = this.attributes.indexOf(key);
    if (index < 0) {
      throw Error(`expected DataSet to contain attribute "${key}"`);
    }
    return index;
  }

  private getDistinctValues (attribute: string): string[] {
    const index = this.getAttributeIndex(attribute);
    return [...new Set<string>(this.records.map(record => record[index]))];
  }

  private getOccurrences (attribute:string, value: string): number {
    const index = this.getAttributeIndex(attribute);
    return this.records.reduce((count, record) => {
      return record[index] === value
        ? ++count
        : count;
    }, 0);
  }

  private getProbability (attribute: string, value: string): number {
    const p = this.getOccurrences(attribute, value) / this.records.length;
    return p * Math.log2(p);
  }

  public getEntropy (attribute: string): number {
    const values = this.getDistinctValues(attribute);
    return values.reduce((entropy, value) => {
      return entropy - this.getProbability(attribute, value);
    }, 0);
  }

  // public getInformationGain (to: string, of: string): number {
  //   const classAttributeEntropy = this.getEntropy(to)
  //
  //
  //   // Gain(Decision, Wind) =
  //   //    Entropy(Decision) –
  //   //    [ p(Decision|Wind=Weak) . Entropy(Decision|Wind=Weak) ] –
  //   //    [ p(Decision|Wind=Strong) . Entropy(Decision|Wind=Strong) ]
  //   // = 0.940 – [ (8/14) . 0.811 ] – [ (6/14). 1]
  //   // = 0.048
  //   return 0;
  // }

  public static fromFilePath (filepath: string): DataSet {
    const [attributes, ...records] = DataSetLoader.load(filepath);
    return new DataSet(attributes, ...records);
  }
}
