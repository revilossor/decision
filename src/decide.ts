#!/usr/bin/env node

import { DecisionTree } from './decision-tree';
import getopt from 'node-getopt';
import Enquirer from 'enquirer';

const { options } = getopt.create([
  ['f', 'filepath=PATH', 'The path to the data set .csv'],
  ['d', 'decision=ATTRIBUTE', 'The attribute we are making a decision for', 'decision'],
  ['i', 'ignore=ATTRIBUTE,ATTRIBUTE,...', 'Attributes not to consider', '']
])
  .bindHelp()
  .parseSystem();

if (!options.filepath) {
  console.error('you must pass a path to a data .csv file (see --help)');
  process.exit(1);
}

const tree = DecisionTree.fromFilePath(
  options.filepath as string,
  {
    classAttribute: (options.decision as string),
    ignoredAttributes: (options.ignore as string).split(',')
  }
);

async function enquire (copy: string, values: string[]):Promise<boolean> {
  const { value } = await Enquirer.prompt({
    type: 'select',
    name: 'value',
    message: copy,
    choices: values
  });
  return value;
}

function say (message: string) {
  console.log(message);
}

async function askQuestion (node): Promise<void> {
  const attribute = node.children[0].attribute;
  if (attribute === `${options.decision}`.toLowerCase()) {
    return say(`I think the answer is "${node.children[0].value.join(' or ')}"`);
  }
  const values = node.children.map(({ value }) => value);
  const result = await enquire(
    `How would you describe the "${attribute}"?`,
    [...values, 'none of the above']
  );
  const next = node.children.find(child => child.value === result);
  if (next) {
    return askQuestion(next);
  }
  return say('I dont know...');
}

askQuestion(tree).then(() => {
  process.exit(0);
}).catch(error => {
  console.error(error);
  process.exit(1);
});
