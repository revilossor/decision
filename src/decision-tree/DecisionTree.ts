import { DataSet } from './DataSet';
import { Node } from './Node';

export class QuestionNode extends Node {
  public constructor (public attribute:string, public value:string) {
    super(`${attribute}==${value}`);
  }
}
export class AnswerNode extends Node {
  public constructor (public attribute:string, public value:string) {
    super(`${attribute}=${value}`);
  }
}

export interface DecisionTreeParams {
  classAttribute: string,
  ignoredAttributes?: string[]
}

export class DecisionTree extends Node {
  private constructor (private data: DataSet, private params: DecisionTreeParams) {
    super(`?${params.classAttribute}/${params.ignoredAttributes}`);
    if (!data.has(params.classAttribute)) {
      throw Error(`expected the class attribute '${params.classAttribute}' to be in the data`);
    }
    // TODO if ignored not present, throw
    this.split(data, this);
  }

  // TODO distinguish homogenous results with > 1 member, return the set?
  private split (data: DataSet, tree: Node) {
    const entropy = data.getEntropy(this.params.classAttribute);
    if (entropy === 0) {
      const decision = data.getDistinctValues(this.params.classAttribute).pop();
      tree.addChild(new AnswerNode(
        this.params.classAttribute.toLowerCase(),
        decision
      ));
    } else {
      const attribute = data.getMostInformative(
        this.params.classAttribute,
        this.params.ignoredAttributes
      );
      const values = data.getDistinctValues(attribute);
      values.forEach(value => {
        const subset = data.subset(attribute, value);
        const node = new QuestionNode(
          attribute,
          value
        );
        tree.addChild(node);
        this.split(subset, node);
      });
    }
  }

  public static fromFilePath (filepath: string, params: DecisionTreeParams): DecisionTree {
    return new DecisionTree(
      DataSet.fromFilePath(filepath),
      params
    );
  }
}
