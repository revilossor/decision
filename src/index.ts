import { DecisionTree } from './decision-tree';

const tree = DecisionTree.fromFilePath(
  './data/playtennis.csv',
  {
    classAttribute: 'Decision',
    ignoredAttributes: ['Day']
  }
);

console.log(JSON.stringify(tree, null, 2));
