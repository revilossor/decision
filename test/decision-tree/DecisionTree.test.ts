import { DecisionTree } from '../../src/decision-tree'; // import from here to check exposed correctly
import { Node } from '../../src/decision-tree/Node';

describe('When I get a DecisionTree instance from a file path', () => {
  describe('And the class attribute is not in the data', () => {
    it('Then an error is thrown', () => {
      const params = {
        classAttribute: 'themoon'
      };
      const expected = Error(`expected the class attribute '${params.classAttribute}' to be in the data`);
      expect(() => DecisionTree.fromFilePath(
        './test/fixtures/playtennis.csv',
        params
      )).toThrow(expected);
    });
  });
  describe('And I choose a class attribute that is in the data', () => {
    let tree:DecisionTree;

    beforeEach(() => {
      tree = DecisionTree.fromFilePath(
        './test/fixtures/playmusic.csv',
        { classAttribute: 'Artist' }
      );
    });

    it('Then the tree is an instance of a Node', () => {
      expect(tree).toBeInstanceOf(Node);
    });

    it('And the correct tree is computed for the data set', () => {
      expect(JSON.stringify(tree)).toMatchSnapshot();
    });
  });
});
