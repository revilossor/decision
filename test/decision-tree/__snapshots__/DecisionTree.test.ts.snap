// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I get a DecisionTree instance from a file path And I choose a class attribute that is in the data And the correct tree is computed for the data set 1`] = `"{\\"value\\":\\"?Artist/undefined\\",\\"children\\":[{\\"value\\":\\"true\\",\\"children\\":[{\\"value\\":\\"true\\",\\"children\\":[{\\"value\\":\\"motorhead\\",\\"children\\":[],\\"attribute\\":\\"artist\\"}],\\"attribute\\":\\"loud\\"},{\\"value\\":\\"false\\",\\"children\\":[{\\"value\\":\\"bela fleck\\",\\"children\\":[],\\"attribute\\":\\"artist\\"}],\\"attribute\\":\\"loud\\"}],\\"attribute\\":\\"fast\\"},{\\"value\\":\\"false\\",\\"children\\":[{\\"value\\":\\"max richter\\",\\"children\\":[],\\"attribute\\":\\"artist\\"}],\\"attribute\\":\\"fast\\"}],\\"data\\":{\\"attributes\\":[\\"fast\\",\\"loud\\",\\"artist\\"],\\"records\\":[[\\"true\\",\\"true\\",\\"motorhead\\"],[\\"true\\",\\"false\\",\\"bela fleck\\"],[\\"false\\",\\"false\\",\\"max richter\\"]]},\\"params\\":{\\"classAttribute\\":\\"Artist\\"}}"`;
